lib/wasix/src/state/context_switching.rs

```diff
@@ -0,0 +1,453 @@
+use crate::{
+    WasiError, WasiFunctionEnv,
+    utils::thread_local_executor::{
+        ThreadLocalExecutor, ThreadLocalSpawner, ThreadLocalSpawnerError,
+    },
+};
+use futures::{
+    TryFutureExt,
+    channel::oneshot::{self, Sender},
+};
+use std::{
+    collections::BTreeMap,
+    mem::forget,
+    sync::{
+        Arc, RwLock, Weak,
+        atomic::{AtomicU64, Ordering},
+    },
+};
+use thiserror::Error;
+use tracing::trace;
+use wasmer::{RuntimeError, Store};
+use wasmer_wasix_types::wasi::ExitCode;
+
+/// The context-switching environment represents all state for WASIX context-switching
+/// on a single host thread.
+#[derive(Debug)]
+pub(crate) struct ContextSwitchingEnvironment {
+    inner: Arc<ContextSwitchingEnvironmentInner>,
+}
+
+#[derive(Debug)]
+struct ContextSwitchingEnvironmentInner {
+    /// List of the unblockers for all suspended contexts
+    unblockers: RwLock<BTreeMap<u64, Sender<Result<(), RuntimeError>>>>,
```

Comment:
Why not HashMap?

Reply:
From what I've seen, everybody generally uses HashMaps unless they need ordered access to data... @theduke would have you use an `FnvHashMap` or `AHashMap` here.